File: apps/mobile/shared/Sources/FlagshipCore/DebugAccess.swift
import Foundation
import CryptoKit

/// Owner-authorized debug-access grant (Swift mirror of
/// packages/protocol/src/debugAccess.ts). The phone signs this behind the
/// create biometric when the user turns on "Debug-friendly server" (Advanced
/// mode) and bakes it into the recipe as the UNSIGNED `debugGrant` sibling; the
/// box verifies it against the owner IRK before enabling the debug console
/// user / SSH.
///
/// Canonical bytes (byte-identical to TS + Kotlin, pinned vector):
///   flagship/debug-access/v1|<serverDomain>|<sshAuthorizedKey>|<issuedAt>
public enum DebugAccess {
    public struct Grant: Codable, Equatable, Sendable {
        public let serverDomain: String
        public let sshAuthorizedKey: String
        public let issuedAt: Int64
        public init(serverDomain: String, sshAuthorizedKey: String, issuedAt: Int64) {
            self.serverDomain = serverDomain
            self.sshAuthorizedKey = sshAuthorizedKey
            self.issuedAt = issuedAt
        }
    }

    public static func canonicalBytes(_ g: Grant) -> Data {
        Data("flagship/debug-access/v1|\(g.serverDomain)|\(g.sshAuthorizedKey)|\(g.issuedAt)".utf8)
    }

    /// Sign with the owner IRK; returns the signature hex.
    public static func sign(_ g: Grant, irk: Curve25519.Signing.PrivateKey) throws -> String {
        let sig = try irk.signature(for: canonicalBytes(g))
        return sig.map { String(format: "%02x", $0) }.joined()
    }

    public static func verify(_ g: Grant, signatureHex: String, irkPub: Data) -> Bool {
        guard let sig = dataFromHex(signatureHex),
              let pub = try? Curve25519.Signing.PublicKey(rawRepresentation: irkPub) else { return false }
        return pub.isValidSignature(sig, for: canonicalBytes(g))
    }

    /// The recipe's `debugGrant` sibling: `{grant:{...}, signatureHex}` (JSON).
    /// Baked into the recipe at mint time and consumed verbatim by the box-side
    /// gate (`debugAccessGate.ts`).
    public static func envelopeJSON(_ g: Grant, signatureHex: String) -> String {
        let body: [String: Any] = [
            "grant": ["serverDomain": g.serverDomain,
                      "sshAuthorizedKey": g.sshAuthorizedKey,
                      "issuedAt": g.issuedAt],
            "signatureHex": signatureHex,
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: body),
              let s = String(data: data, encoding: .utf8) else { return "{}" }
        return s
    }

    private static func dataFromHex(_ s: String) -> Data? {
        guard s.count % 2 == 0 else { return nil }
        var out = Data(capacity: s.count / 2)
        var i = s.startIndex
        while i < s.endIndex {
            let n = s.index(i, offsetBy: 2)
            guard let b = UInt8(s[i..<n], radix: 16) else { return nil }
            out.append(b); i = n
        }
        return out
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/DebugAccess.kt
package com.flagshipserver.app.core

import com.google.crypto.tink.subtle.Ed25519Sign
import com.google.crypto.tink.subtle.Ed25519Verify
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Owner-authorized debug-access grant (Kotlin mirror of
 * packages/protocol/src/debugAccess.ts + the iOS DebugAccess). The phone
 * signs this behind biometric when the user approves the burner's "Debug
 * mode" toggle over the live pairing session; the box verifies it against
 * the owner IRK before enabling the debug console user / SSH.
 *
 * Canonical bytes (byte-identical to TS + Swift, pinned vector):
 *   flagship/debug-access/v1|<serverDomain>|<sshAuthorizedKey>|<issuedAt>
 */
object DebugAccess {
    data class Grant(val serverDomain: String, val sshAuthorizedKey: String, val issuedAt: Long)

    fun canonicalBytes(g: Grant): ByteArray =
        "flagship/debug-access/v1|${g.serverDomain}|${g.sshAuthorizedKey}|${g.issuedAt}".toByteArray()

    fun sign(g: Grant, irk: Ed25519Sign): String = HexUtil.encode(irk.sign(canonicalBytes(g)))

    fun verify(g: Grant, signatureHex: String, irkPub: ByteArray): Boolean = try {
        Ed25519Verify(irkPub).verify(HexUtil.decode(signatureHex), canonicalBytes(g))
        true
    } catch (_: Throwable) {
        false
    }

    /** The on-wire `consent-result` payload: {grant:{...}, signatureHex}. */
    fun envelopeJson(g: Grant, signatureHex: String): String {
        val grant: JsonObject = buildJsonObject {
            put("serverDomain", g.serverDomain)
            put("sshAuthorizedKey", g.sshAuthorizedKey)
            put("issuedAt", g.issuedAt)
        }
        return buildJsonObject {
            put("grant", grant)
            put("signatureHex", signatureHex)
        }.toString()
    }
}
